File: ruby/handler.rb
require 'json'
require 'etc'

# Module initialization timestamp
MODULE_INIT_TIME = Process.clock_gettime(Process::CLOCK_MONOTONIC)
MODULE_INIT_DATE = Time.now.utc.iso8601

# Cold start detection with more robust logic
$container_start_time = MODULE_INIT_TIME
$invocation_count = 0
$last_invocation_time = 0

# Cold start threshold: 5 minutes of inactivity
COLD_START_THRESHOLD_MS = 5 * 60 * 1000

def detect_cold_start
  now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  time_since_last_invocation = (now - $last_invocation_time) * 1000 # Convert to ms
  
  # First invocation or long period of inactivity
  is_cold = $invocation_count == 0 || time_since_last_invocation > COLD_START_THRESHOLD_MS
  
  if is_cold && $invocation_count > 0
    # Container was reused after inactivity - reset metrics
    $container_start_time = now
  end
  
  is_cold
end

def get_memory_usage
  # Get memory usage from /proc/self/status if available (Linux)
  if File.exist?('/proc/self/status')
    status = File.read('/proc/self/status')
    vm_rss = status[/VmRSS:\s+(\d+)\s+kB/, 1]
    vm_size = status[/VmSize:\s+(\d+)\s+kB/, 1]
    
    {
      rss: vm_rss ? (vm_rss.to_f / 1024).round(2) : 0, # MB
      vms: vm_size ? (vm_size.to_f / 1024).round(2) : 0 # MB
    }
  else
    # Fallback for non-Linux systems
    {
      rss: 0,
      vms: 0
    }
  end
end

def get_system_info
  {
    cpuCount: Etc.nprocessors,
    platform: RbConfig::CONFIG['host_os'],
    arch: RbConfig::CONFIG['host_cpu'],
    rubyVersion: RUBY_VERSION
  }
end

def hello(event:, context:)
  handler_start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  request_id = context&.aws_request_id || 'local-test'
  is_cold = detect_cold_start
  
  $invocation_count += 1
  $last_invocation_time = handler_start_time
  
  # Simulate some work to measure handler execution time
  work_start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  
  # Get system information
  memory_usage = get_memory_usage
  system_info = get_system_info
  
  work_end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  handler_end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  
  # Calculate precise timings (all in milliseconds with 2 decimal precision)
  timings = {
    moduleInitDurationMs: ((handler_start_time - MODULE_INIT_TIME) * 1000).round(2),
    handlerDurationMs: ((handler_end_time - handler_start_time) * 1000).round(2),
    workDurationMs: ((work_end_time - work_start_time) * 1000).round(2),
    totalContainerAgeMs: ((handler_start_time - $container_start_time) * 1000).round(2)
  }
  
  # Standardized response format
  response = {
    # Basic info
    message: "HELLO WORLD",
    runtime: "ruby3.2",
    timestamp: Time.now.utc.iso8601,
    requestId: request_id,
    
    # Cold start detection
    coldStart: is_cold,
    invocationCount: $invocation_count,
    
    # Timing metrics (all in milliseconds with 2 decimal precision)
    timings: timings,
    
    # System metrics
    system: {
      memoryUsage: memory_usage,
      **system_info
    },
    
    # Container lifecycle
    container: {
      initTimestamp: MODULE_INIT_DATE,
      ageMs: timings[:totalContainerAgeMs]
    }
  }
  
  # Structured logging
  log_entry = {
    level: "INFO",
    timestamp: response[:timestamp],
    requestId: request_id,
    coldStart: is_cold,
    invocationCount: $invocation_count,
    timings: timings,
    memoryUsage: memory_usage[:rss],
    message: "Lambda invocation completed"
  }
  puts JSON.generate(log_entry)
  
  {
    statusCode: 200,
    headers: {
      "Content-Type" => "application/json",
      "X-Cold-Start" => is_cold.to_s,
      "X-Invocation-Count" => $invocation_count.to_s,
      "X-Handler-Duration-Ms" => timings[:handlerDurationMs].to_s,
      "X-Runtime" => "ruby3.2"
    },
    body: JSON.pretty_generate(response)
  }
end

File: java/src/main/java/com/example/Handler.java
package com.example;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPResponse;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.MemoryUsage;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

// Java 21 Records for structured data
record TimingMetrics(
    double moduleInitDurationMs,
    double handlerDurationMs,
    double workDurationMs,
    double totalContainerAgeMs
) {}

record MemoryMetrics(
    double heapUsed,
    double heapMax,
    double nonHeapUsed
) {}

record SystemInfo(
    int cpuCount,
    String platform,
    String arch,
    String javaVersion
) {}

record ContainerInfo(
    String initTimestamp,
    double ageMs
) {}

public class Handler implements RequestHandler<APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse> {

    // Module initialization timestamp
    private static final long MODULE_INIT_TIME = System.nanoTime();
    private static final String MODULE_INIT_DATE = Instant.now().toString();
    
    // Cold start detection with more robust logic
    private static long containerStartTime = MODULE_INIT_TIME;
    private static int invocationCount = 0;
    private static long lastInvocationTime = 0;
    
    // Cold start threshold: 5 minutes of inactivity
    private static final long COLD_START_THRESHOLD_MS = 5 * 60 * 1000;
    
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final MemoryMXBean memoryBean = ManagementFactory.getMemoryMXBean();
    
    private static boolean detectColdStart() {
        long now = System.nanoTime();
        long timeSinceLastInvocation = (now - lastInvocationTime) / 1_000_000; // Convert to ms
        
        // First invocation or long period of inactivity
        boolean isCold = invocationCount == 0 || timeSinceLastInvocation > COLD_START_THRESHOLD_MS;
        
        if (isCold && invocationCount > 0) {
            // Container was reused after inactivity - reset metrics
            containerStartTime = now;
        }
        
        return isCold;
    }
    
    private static MemoryMetrics getMemoryUsage() {
        MemoryUsage heapUsage = memoryBean.getHeapMemoryUsage();
        MemoryUsage nonHeapUsage = memoryBean.getNonHeapMemoryUsage();
        
        return new MemoryMetrics(
            Math.round(heapUsage.getUsed() / 1024.0 / 1024.0 * 100.0) / 100.0, // MB
            Math.round(heapUsage.getMax() / 1024.0 / 1024.0 * 100.0) / 100.0, // MB
            Math.round(nonHeapUsage.getUsed() / 1024.0 / 1024.0 * 100.0) / 100.0 // MB
        );
    }
    
    private static SystemInfo getSystemInfo() {
        return new SystemInfo(
            Runtime.getRuntime().availableProcessors(),
            System.getProperty("os.name"),
            System.getProperty("os.arch"),
            System.getProperty("java.version")
        );
    }

    @Override
    public APIGatewayV2HTTPResponse handleRequest(APIGatewayV2HTTPEvent event, Context context) {
        long handlerStartTime = System.nanoTime();
        String requestId = context != null ? context.getAwsRequestId() : "local-test";
        boolean isCold = detectColdStart();
        
        invocationCount++;
        lastInvocationTime = handlerStartTime;
        
        // Simulate some work to measure handler execution time
        long workStartTime = System.nanoTime();
        
        // Get system information
        MemoryMetrics memoryUsage = getMemoryUsage();
        SystemInfo systemInfo = getSystemInfo();
        
        long workEndTime = System.nanoTime();
        long handlerEndTime = System.nanoTime();
        
        // Calculate precise timings using Java 21 records
        TimingMetrics timings = new TimingMetrics(
            Math.round((handlerStartTime - MODULE_INIT_TIME) / 1_000_000.0 * 100.0) / 100.0,
            Math.round((handlerEndTime - handlerStartTime) / 1_000_000.0 * 100.0) / 100.0,
            Math.round((workEndTime - workStartTime) / 1_000_000.0 * 100.0) / 100.0,
            Math.round((handlerStartTime - containerStartTime) / 1_000_000.0 * 100.0) / 100.0
        );
        
        // Container lifecycle info using record
        ContainerInfo containerInfo = new ContainerInfo(
            MODULE_INIT_DATE,
            timings.totalContainerAgeMs()
        );
        
        // System metrics combining memory and system info
        Map<String, Object> system = Map.of(
            "memoryUsage", Map.of(
                "heapUsed", memoryUsage.heapUsed(),
                "heapMax", memoryUsage.heapMax(),
                "nonHeapUsed", memoryUsage.nonHeapUsed()
            ),
            "cpuCount", systemInfo.cpuCount(),
            "platform", systemInfo.platform(),
            "arch", systemInfo.arch(),
            "javaVersion", systemInfo.javaVersion()
        );
        
        // Standardized response format using Java 21 features
        var timestamp = Instant.now().toString();
        Map<String, Object> response = Map.of(
            "message", "HELLO WORLD",
            "runtime", "java21",
            "timestamp", timestamp,
            "requestId", requestId,
            "coldStart", isCold,
            "invocationCount", invocationCount,
            "timings", Map.of(
                "moduleInitDurationMs", timings.moduleInitDurationMs(),
                "handlerDurationMs", timings.handlerDurationMs(),
                "workDurationMs", timings.workDurationMs(),
                "totalContainerAgeMs", timings.totalContainerAgeMs()
            ),
            "system", system,
            "container", Map.of(
                "initTimestamp", containerInfo.initTimestamp(),
                "ageMs", containerInfo.ageMs()
            )
        );
        
        // Structured logging using immutable maps
        Map<String, Object> logEntry = Map.of(
            "level", "INFO",
            "timestamp", timestamp,
            "requestId", requestId,
            "coldStart", isCold,
            "invocationCount", invocationCount,
            "timings", Map.of(
                "moduleInitDurationMs", timings.moduleInitDurationMs(),
                "handlerDurationMs", timings.handlerDurationMs(),
                "workDurationMs", timings.workDurationMs(),
                "totalContainerAgeMs", timings.totalContainerAgeMs()
            ),
            "memoryUsage", memoryUsage.heapUsed(),
            "message", "Lambda invocation completed"
        );
        
        try {
            System.out.println(objectMapper.writeValueAsString(logEntry));
        } catch (Exception e) {
            System.err.println("Failed to log: " + e.getMessage());
        }
        
        // Response headers using immutable Map
        Map<String, String> headers = Map.of(
            "Content-Type", "application/json",
            "X-Cold-Start", String.valueOf(isCold),
            "X-Invocation-Count", String.valueOf(invocationCount),
            "X-Handler-Duration-Ms", String.valueOf(timings.handlerDurationMs()),
            "X-Runtime", "java21"
        );
        
        try {
            String responseBody = objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(response);
            return APIGatewayV2HTTPResponse.builder()
                    .withStatusCode(200)
                    .withHeaders(headers)
                    .withBody(responseBody)
                    .build();
        } catch (Exception e) {
            System.err.println("Failed to serialize response: " + e.getMessage());
            return APIGatewayV2HTTPResponse.builder()
                    .withStatusCode(500)
                    .withHeaders(Map.of("Content-Type", "application/json"))
                    .withBody("{\"error\":\"Internal server error\"}")
                    .build();
        }
    }
}
